### PR TITLE
Add security guide to the web page

### PR DESCRIPTION
**Pull request by cmcantalupo**
- Relates to #2120

### CODE REVIEW

File: service/docs/source/security.rst [hunk 1]
```diff
@@ -0,0 +1,525 @@
...
+
+File Usage/Configuration
+^^^^^^^^^^^^^^^^^^^^^^^^
+
+As GEOPM reads and writes configuration files to disk, it is important
+to validate that the file usage is done as securely as possible. This
+includes verifying an input file is not a symbolic link to an
+unintended resource, verifying that all temporary files are cleaned up
+properly, and that the temporary files and directories that are used
+to not have excessive permissions.
```
Comment (avilcheslopez) [inline service/docs/source/security.rst]:
```suggestion
do not have excessive permissions.
```

File: service/docs/source/security.rst [hunk 2]
```diff
@@ -0,0 +1,525 @@
...
+this temporary file. Once writing is complete, the temporary file is
+renamed to the desired path while preserving the 0o600
+permissions. This rename is atomic, so it is not possible for files to
+exist with partial/corrupt data. Any existing file at the desired
+location will be overwritten.
+
+When reading files, the path is first checked to ensure it is secure.
+This includes asserting that the path exists, it is not a link, it is
+not a directory, and it is a regular file. After the path is verified,
+it a file descriptor is opened for the path and then checked
```
Comment (avilcheslopez) [inline service/docs/source/security.rst]: Minor note: check grammar around "it".

File: service/docs/source/security.rst [hunk 3]
```diff
@@ -0,0 +1,525 @@
...
+
+The GEOPM Service has the capability of writing to the logging service
+provided by ``dasbus``. Most commonly this is the logging capabilities
+provided by systemd, and are accessible via the journalctl command
+(e.g.  journalctl -u geopm) or through inspection of
+``/var/log/messages`` or similar ways of viewing the ``syslog`` which
+may depend on how the system is configured.
+
+The GEOPM Service will log any error conditions that arise from
+attempting to setup or use configuration files stored in a secure
```
Comment (avilcheslopez) [inline service/docs/source/security.rst]:
```suggestion
attempting to set up or use configuration files stored in a secure
```

File: service/docs/source/security.rst [hunk 4]
```diff
@@ -0,0 +1,525 @@
...
+how the surface is secured against this threat.
+
+
+Malicious input or private output
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Each GEOPM Service interface that may be accessed by an unprivileged
+user is a threat vector.  All input recieved from the user is
+validated to ensure no malicious or malformed data is used in any way
+that my result in a compromised or misconfigured system. All output
```
Comment (avilcheslopez) [inline service/docs/source/security.rst]:
```suggestion
that may result in a compromised or misconfigured system. All output
```

File: service/docs/source/security.rst [hunk 5]
```diff
@@ -0,0 +1,525 @@
...
+from these interfaces is vetted to ensure that the service does not
+emit private, malicious or malformed data.
+
+The GEOPM Service provides two interfaces for end users.  One is the
+``io.github.geopm`` DBus interface published through systemd.  The
+other is the batch server interface which may be created with a
+request to through GEOPM DBus interface.  The batch server is accessed
```
Comment (avilcheslopez) [inline service/docs/source/security.rst]: Minor issue: "to through"

**Review by avilcheslopez — COMMENTED** [review]
Looks good to me!

File: service/docs/source/security.rst [hunk 6]
```diff
@@ -0,0 +1,525 @@
...
+High Level Security Objectives
+------------------------------
+
+The GEOPM Service provides a secure gateway to privileged hardware
+features. This section describes the security expectations of the
+system administrator who deploys and configures the GEOPM Service as
+described in the GEOPM documentation.
+
+
+Fine-grained access management for hardware features
```
Comment (bgeltz) [inline service/docs/source/security.rst:21]: You're mixing upper/lower case for these section titles.  Do you want it like this instead:
```suggestion
Fine-Grained Access Management for Hardware Features
```

File: service/docs/source/security.rst [hunk 7]
```diff
@@ -0,0 +1,525 @@
...
+`capabilities(7) <https://man7.org/linux/man-pages/man7/capabilities.7.html>`__
+requirements.
+
+Granting Linux capabilites to a helper command line tool or adding a
+command line tool to the
+`sudoers(5) <https://man7.org/linux/man-pages/man5/sudoers.5.html>`__
+configuration is a common way to filter which secure features are
+enabled.  These command line tools may also require independent
+security review and typically do not provide a high performance
+programatic solution.
```
Comment (bgeltz) [inline service/docs/source/security.rst]:
```suggestion
programmatic solution.
```

File: service/docs/source/security.rst [hunk 8]
```diff
@@ -0,0 +1,525 @@
...
+interface.
+
+Each hardware feature that may be read is referred to as a "signal"
+and mapped to a specific string name. Each hardware feature that may
+be written is referred to as a "control" and is mapped to a specific
+string name. These signals and controls are represented as double
+precision floating point numbers in SI units. Requests to GEOPM
+interfaces to read a signal or write a control are combined with a
+hardware domain and a domain index to identify the specific hardware
+component that the request applies to.
```
Comment (bgeltz) [inline service/docs/source/security.rst]:
```suggestion
component to which the request applies.
```

File: service/docs/source/security.rst [hunk 9]
```diff
@@ -0,0 +1,525 @@
...
+controls until a system administrator grants this access; this follows
+the "Secure by Default" design. All available signals and controls are
+documented, and the description includes the security implications of
+granting access. The ``geopmaccess`` command line tool is used by a
+system administrator to query descriptions of signals and controls
+available on a system as well as managing user access to each feature
+by name.
+
+
+Restore control configuration after a writing session ends
```
Comment (bgeltz) [inline service/docs/source/security.rst:86]:
```suggestion
Restore Control Configuration After a Writing Session Ends
```

File: service/docs/source/security.rst [hunk 10]
```diff
@@ -0,0 +1,525 @@
...
+settings are restored to their previous value. The storage location
+for the saved state is in the directory
+``/var/run/geopm-service/SAVE_FILES``.
+
+Note that all configuration settings are restored regardless of
+whether or not they were adjusted during the session through the
+service.
+
+
+One writing GEOPM session allowed at any time
```
Comment (bgeltz) [inline service/docs/source/security.rst:110]:
```suggestion
One Writing GEOPM Session Allowed at Any Time
```

File: service/docs/source/security.rst [hunk 11]
```diff
@@ -0,0 +1,525 @@
...
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+GEOPM Service enforces that at most one writing user session is
```
Comment (bgeltz) [inline service/docs/source/security.rst]:
```suggestion
The GEOPM Service enforces that at most one writing user session is
```

File: service/docs/source/security.rst [hunk 12]
```diff
@@ -0,0 +1,525 @@
...
+write lock that enables this mutual exclusion is associated with the
+Unix "process session leader" PID of the requesting process. If this
+session leader PID does not refer to an active process, then the lock
+is associated with the requesting process PID. Otherwise, any PID
+within the requesting Unix process session is enabled to write through
+the GEOPM Service for the duration of the process session leader PID
+lifetime.
+
+
+Provide a low latency interface
```
Comment (bgeltz) [inline service/docs/source/security.rst:125]:
```suggestion
Provide a Low Latency Interface
```

File: service/docs/source/security.rst [hunk 13]
```diff
@@ -0,0 +1,525 @@
...
+by
+`systemd(1) <https://man7.org/linux/man-pages/man1/systemd.1.html>`__.
+The user configures the batch server with an array of requests through
+DBus. The access permissions are checked when the call to start the
+server is made, and the server is only forked if the client has
+permission for all requests in the array. The batch server creates one
+or two inter-process shared memory regions to interact with the
+client: one region provides a user interface to read signals, and the
+other if requested provides a user interface to write
+controls. Requests to read, write or quit are made through a FIFO
```
Comment (bgeltz) [inline service/docs/source/security.rst]:
```suggestion
controls. Requests to read, write, or quit are made through a FIFO
```

I can't remember if we do serial commas or not.  I think we waffle back and forth on that and one or two spaces after the period.

File: service/docs/source/security.rst [hunk 14]
```diff
@@ -0,0 +1,525 @@
...
+channel connecting the client to the batch server, and a second FIFO
+is used to communicate to the client when requests have been
+completed. These FIFOs are opened in ``/tmp`` which by default systemd
+creates as a `tmpfs(5)
+<https://man7.org/linux/man-pages/man5/tmpfs.5.html>`__.
+
+
+Secure restart of the ``geopmd`` daemon process
```
Comment (bgeltz) [inline service/docs/source/security.rst:148]:
```suggestion
Secure Restart of the ``geopmd`` Daemon Process
```

File: service/docs/source/security.rst [hunk 15]
```diff
@@ -0,0 +1,525 @@
...
+the user may then interact with this batch server through faster
+mechanisms than DBus provides. In particular, the user interfaces with
+the batch server over inter-process shared memory through the
+``/dev/shm`` device, and sends commands through FIFO special files in
+``/tmp``.
+
+The DBus interface provides a layer of security that is leveraged
+throughout Linux services to verify the user identity of requests made
+to daemon processes. The GEOPM Service relies on the systemd DBus
+interface to provide the PID, UID, and GID of the requesting client.
```
Comment (bgeltz) [inline service/docs/source/security.rst:181]: SERIAL COMMA! xD

File: service/docs/source/security.rst [hunk 16]
```diff
@@ -0,0 +1,525 @@
...
+The GEOPM Service provides the ``geopmaccess`` command line interface
+for system administrators to manage access to hardware features. This
+interface is expected to be a reliable and secure mechanism for
+managing access rights of users to the assets described in this
+section.  Maintaining user privacy such that client interactions with
+the GEOPM service are not observable to other users is also a security
+priority.
+
+
+Model Specific Register device driver
```
Comment (bgeltz) [inline service/docs/source/security.rst:229]:
```suggestion
Model Specific Register Device Driver
```

File: service/docs/source/security.rst [hunk 17]
```diff
@@ -0,0 +1,525 @@
...
+administrator to control precisely which subset of features available
+through the MSR driver may be accessed. Additionally, the GEOPM
+Service does not allow persistent changes to the MSR driver. For these
+reasons a system administrator may wish to provide MSR access through
+the GEOPM Service to processes that do not have the CAP_SYS_RAWIO
+Linux
+`capability <https://man7.org/linux/man-pages/man7/capabilities.7.html>`__.
+
+
+Intel Speed Select device driver
```
Comment (bgeltz) [inline service/docs/source/security.rst:259]:
```suggestion
Intel Speed Select Device Driver
```

File: service/docs/source/security.rst [hunk 18]
```diff
@@ -0,0 +1,525 @@
...
+`SST-CP <https://www.kernel.org/doc/html/latest/admin-guide/pm/intel-speed-select.html#intel-r-speed-select-technology-core-power-intel-r-sst-cp>`__
+and
+`SST-TF <https://www.kernel.org/doc/html/latest/admin-guide/pm/intel-speed-select.html#intel-r-speed-select-technology-turbo-frequency-intel-r-sst-tf>`__
+features. Using the ``isst_interface`` device driver requires the
+Linux
+`capability <https://man7.org/linux/man-pages/man7/capabilities.7.html>`__
+of CAP_SYS_ADMIN because changes may impact system performance for
+other users of the system. The ISST interface may also be used to
+change the hardware characteristics reported by the Linux kernel, such
+as number of cores, base frequency, and which turbo frequencies are
```
Comment (bgeltz) [inline service/docs/source/security.rst:279]: SERIAL COMMA! xD

File: service/docs/source/security.rst [hunk 19]
```diff
@@ -0,0 +1,525 @@
...
+achievable.
+
+
+LevelZero sysman library interface
```
Comment (bgeltz) [inline service/docs/source/security.rst:283]:
```suggestion
LevelZero Sysman Library Interface
```

File: service/docs/source/security.rst [hunk 20]
```diff
@@ -0,0 +1,525 @@
...
+The LevelZero sysman library interface enables users to monitor and
+control Intel accelerator devices. These controls and monitors include
+setting bounds on GPU operating frequency and reading performance
+counters from GPU devices. Access to the LevelZero sysman interface is
+restricted because it provides ability to modify the performance of
+the system and direct access to hardware metrics that reflect user
+activity.
+
+
+Nvidia NVML device management library interface
```
Comment (bgeltz) [inline service/docs/source/security.rst:295]:
```suggestion
Nvidia NVML Device Management Library Interface
```

File: service/docs/source/security.rst [hunk 21]
```diff
@@ -0,0 +1,525 @@
...
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+The NVML library interface enables users to monitor and control Nvidia
+GPU devices. These controls and monitors include setting bounds on GPU
```
Comment (bgeltz) [inline service/docs/source/security.rst:299]:
```suggestion
GPU devices. These controls and signals include setting bounds on GPU
```

File: service/docs/source/security.rst [hunk 22]
```diff
@@ -0,0 +1,525 @@
...
+operating frequency and reading performance counters from GPU devices.
+Access to some of the interfaces of the NVML library are restricted
+and some of those restrictions may be relaxed based on settings
+enabled by a system administrator.
+
+
+Service user data
```
Comment (bgeltz) [inline service/docs/source/security.rst:306]:
```suggestion
Service User Data
```

File: service/docs/source/security.rst [hunk 23]
```diff
@@ -0,0 +1,525 @@
...
+^^^^^^^^^^^^^^^^^
+
+Interactions with the GEOPM Service by each client are private
+information and an asset that the GEOPM Service is designed to
+protect.  Unprivileged users of the system should not be able to
+observe the calls, inputs or outputs to the GEOPM Service by other
```
Comment (bgeltz) [inline service/docs/source/security.rst]:
```suggestion
observe the calls, inputs, or outputs to the GEOPM Service by other
```

File: service/docs/source/security.rst [hunk 24]
```diff
@@ -0,0 +1,525 @@
...
+
+
+Attack Surface
+--------------
+
+This section describes the interfaces that must be protected to ensure
+the security requirements of the GEOPM Service.
+
+
+System files
```
Comment (bgeltz) [inline service/docs/source/security.rst:323]:
```suggestion
System Files
```

File: service/docs/source/security.rst [hunk 25]
```diff
@@ -0,0 +1,525 @@
...
+controlling access permissions are located in the
+``/etc/geopm-service`` directory. The state required to support active
+user sessions is stored in ``/var/run/geopm-service``. Protecting
+these files is paramount to the integrity of the GEOPM Service
+security model.  In general these files have root access permissions
+only, and are modified by calling into GEOPM Service interfaces, or by
+running GEOPM Service command line tools like ``geopmaccess``.
+
+
+Inter-process shared memory
```
Comment (bgeltz) [inline service/docs/source/security.rst:337]:
```suggestion
Inter-Process Shared Memory
```

File: service/docs/source/security.rst [hunk 26]
```diff
@@ -0,0 +1,525 @@
...
+configurations with a user processes. For each batch session opened by
+a user process there are one or two shared memory regions opened for
+communication. Protecting the integrity of these shared system
+resources is a critical part of our security model. Access to these
+shared memory regions by a user other than the client can result in
+escalation of privileges, and subversion of the administrative access
+lists.
+
+
+FIFO special files
```
Comment (bgeltz) [inline service/docs/source/security.rst:351]:
```suggestion
FIFO Special Files
```

File: service/docs/source/security.rst [hunk 27]
```diff
@@ -0,0 +1,525 @@
...
+special files created in ``/tmp``, the temporary file system, to
+support the batch server features of the GEOPM Service. Notifications
+sent between the client and server about updates to data in
+inter-process shared memory go across the FIFOs which serve as a
+synchronization mechanism.  Adversarial access to these files could
+cause deadlocks in the batch server or client process or enable
+information about the client session requests to be observed.
+
+
+Systemd DBus interface
```
Comment (bgeltz) [inline service/docs/source/security.rst:364]:
```suggestion
Systemd DBus Interface
```

File: service/docs/source/security.rst [hunk 28]
```diff
@@ -0,0 +1,525 @@
...
+`sd-bus(3) <https://man7.org/linux/man-pages/man3/sd-bus.3.html>`__
+interface.  The GEOPM Service relies on these standard Linux tools to
+provide a trusted interface and a secured attack surface.
+
+
+Logging
+^^^^^^^
+
+GEOPM's implementation ensures that GEOPM has sufficient logging to
+provide traceability to sysadmins about user interactions. This
```
Comment (bgeltz) [inline service/docs/source/security.rst]:
```suggestion
provide traceability to system administrators about user interactions. This
```

This is the first occurrence of "sysadmin".

File: service/docs/source/security.rst [hunk 29]
```diff
@@ -0,0 +1,525 @@
...
+
+This section enumerates the mechanisms that an adversary may use in an
+attempt to breach the attack surface to subvert the security
+guarantees of the GEOPM Service. Each threat is split out into its own
+subsection.  The threat is described in the context of how the
+mechanism might be used against the GEOPM Service attack surface, and
+how the surface is secured against this threat.
+
+
+Malicious input or private output
```
Comment (bgeltz) [inline service/docs/source/security.rst:422]:
```suggestion
Malicious Input or Private Output
```

File: service/docs/source/security.rst [hunk 30]
```diff
@@ -0,0 +1,525 @@
...
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Each GEOPM Service interface that may be accessed by an unprivileged
+user is a threat vector.  All input recieved from the user is
+validated to ensure no malicious or malformed data is used in any way
+that my result in a compromised or misconfigured system. All output
+from these interfaces is vetted to ensure that the service does not
+emit private, malicious or malformed data.
```
Comment (bgeltz) [inline service/docs/source/security.rst]:
```suggestion
emit private, malicious, or malformed data.
```

**Review by bgeltz — CHANGES_REQUESTED** [review]

File: service/docs/source/security.rst (hunk 6 above)
Comment (cmcantalupo) [reply service/docs/source/security.rst:21]: The intent was ^^^^ titles (which are longer) are not capitalized. Will fix.

**Review by bgeltz — APPROVED** [review]

**Pull request merged**